doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+from the most recent snapshot to the current snapshot.
+
+Git knows that people often work on multiple changes in multiple files all at
+once, but that ultimately they should separate those changes into sets of
+related changes that are grouped together based on common goals (so that it's
+easier for their colleagues to understand and review the changes). For example,
+you might want to group all the code changes together in one commit, put new
+unit tests in another commit, and changes to the documentation in a third
+commit.  Git makes this easy(ish) with something called the `stage`_ (or
+*staging area*). After you've made some changes to the codebase, you'll have
+what git calls "unstaged changes", which will show up with the `status`_
+command:
+
+.. code-block:: console
+
+    $ # see what state the local copy of the codebase is in
+    $ git status
+
+Those unstaged changes can be `added`_ to the stage one by one, by either
+adding a whole file's worth of changes, or by adding only certain lines
+interactively:
+
+.. code-block:: console
+
+    $ # add a whole file's worth of changes
+    $ # (same command works to add a completely new file):
+    $ git add mne/some_file.py
+    $ # enter interactive staging mode, to add only portions of the file:
+    $ git add -p mne/viz/some_other_file.py
+
+Once you've collected all the related changes together on the stage, the ``git
+status`` command will now refer to them as "changes staged for commit". You can
+commit them to the current branch with the `commit`_ command. If you just type
+``git commit`` by itself, git will open the text editor you configured it to
+use so that you can write a *commit message* — a short description of the
+changes you've grouped together in this commit. You can bypass the text editor
+by passing a commit message on the command line with the ``-m`` flag. For
+example, if your first commit adds a new feature, your commit message might be:
+
+.. code-block:: console
+
+    $ git commit -m 'ENH: adds feature X to the Epochs class'
+
+Once you've made the commit, the stage is now empty, and you can repeat the
+cycle, adding the unit tests and documentation changes:
+
+.. code-block:: console
+
+    $ git add mne/tests/some_testing_file.py
+    $ git commit -m 'add test of new feature X of the Epochs class'
+    $ git add -p mne/some_file.py mne/viz/some_other_file.py
+    $ git commit -m 'DOC: update Epochs and BaseEpochs docstrings'
+    $ git add tutorials/new_tutorial_file.py
+    $ git commit -m 'DOC: adds new tutorial about feature X'
+
+When you're done, it's time to run the test suite to make sure your changes
+haven't broken any existing functionality, and to make sure your new test
+covers the lines of code you've added (see :ref:`run-tests` and
+:ref:`build-docs`, below). Once everything looks good, it's time to push your
+changes to your fork:
+
+.. code-block:: console
+
+    $ git push origin new-feature-x
+
```

Comment:
What I find really useful (especially when pushing several times) is `git push -u origin new-feature-x` for the first push, than you only have to type `git push` afterwards. Or, even more convenient, configure your git by `git config --global push.default current`, then the first push of a new branch becomes simply `git push -u`.

Reply:
hmm.  I like the suggestion, but I think maybe this is another case where the extra explanation / length is not worth what we gain here.  Another candidate for an "advanced git / troubleshooting" page?

Reply:
Agreed we should push most tricks to another doc. Let's show people one good way to do things here, and the advanced people can learn from each other somewhere else

Reply:
Sure. (I just learned it with the -u, I event did not know it also works without it.) Don't you think at least including the -u for a first push makes things easier?
E.g.:
```(bash)
git push -u origin new-feature-x  # for the first time you push a new branch
# or:
git push  # for all following pushes of the same branch
```

Reply:
punted into #6266